test/unit/advertisement-operator/controller_test.go

```diff
@@ -0,0 +1,126 @@
+	}
+}
+
+func TestCheckAdvertisementWithAutoAccept(t *testing.T) {
+	r := createReconciler(0, 10, true)
+
+	// ACCEPT TEST
+	// given a configuration with max 10 Advertisements, create 10 Advertisements and check that they are all accepted
+	for i := 0; i < 10; i++ {
+		adv := createAdvertisement()
+		r.CheckAdvertisement(&adv)
+		assert.Equal(t, "ACCEPTED", adv.Status.AdvertisementStatus)
+	}
+	// check that the Adv counter has been incremented
+	assert.Equal(t, int32(10), r.AcceptedAdvNum)
+
+	// REFUSE TEST
+	// create 10 more Advertisements and check that they are all refused, since the maximum has been reached
+	for i := 0; i < 10; i++ {
+		adv := createAdvertisement()
+		r.CheckAdvertisement(&adv)
+		assert.Equal(t, "REFUSED", adv.Status.AdvertisementStatus)
+	}
+	// check that the Adv counter has not been modified
+	assert.Equal(t, int32(10), r.AcceptedAdvNum)
+}
+
+func TestCheckAdvertisementWithoutAutoAccept(t *testing.T) {
+	r := createReconciler(0, 10, false)
+
+	// given a configuration with max 10 Advertisements but no AutoAccept, create 10 Advertisements and check that their status is not set
+
+	for i := 0; i < 10; i++ {
+		adv := createAdvertisement()
+		r.CheckAdvertisement(&adv)
+		assert.Equal(t, "", adv.Status.AdvertisementStatus)
+	}
+	// check that the Adv counter has not been incremented
+	assert.Equal(t, int32(0), r.AcceptedAdvNum)
+}
+
+func TestManageConfigUpdate(t *testing.T) {
+	r := createReconciler(0, 10, true)
+	advList := v1.AdvertisementList{
+		Items: []v1.Advertisement{},
+	}
+
+	advCount := 15
+
+	// given a configuration with max 10 Advertisements, create 15 Advertisement: 10 should be accepted and 5 refused
+	for i := 0; i < advCount; i++ {
+		adv := createAdvertisement()
+		r.CheckAdvertisement(&adv)
+		advList.Items = append(advList.Items, adv)
+	}
+
+	// the advList contains 10 accepted and 5 refused Adv
+	// create a new configuration with MaxAcceptableAdv = 15
+	// with the new configuration, check the 5 refused Adv are accepted
+	config := policyv1.ClusterConfig{
+		Spec: policyv1.ClusterConfigSpec{
+			MaxAcceptableAdvertisement: int32(advCount),
+			AutoAccept:                 true,
+		},
+	}
+
+	// TRUE TEST
+	// test the true branch of ManageConfigUpdate
+	err, flag := r.ManageConfigUpdate(&config, &advList)
+	assert.Nil(t, err)
+	assert.True(t, flag)
+	assert.Equal(t, config.Spec, r.ClusterConfig)
+	assert.Equal(t, int32(advCount), r.AcceptedAdvNum)
+	for _, adv := range advList.Items {
+		assert.Equal(t, "ACCEPTED", adv.Status.AdvertisementStatus)
+	}
+
+	// FALSE TEST
+	// apply again the same configuration
+	// we enter in the false branch of ManageConfigUpdate but nothing should change
+	err, flag = r.ManageConfigUpdate(&config, &advList)
+	assert.Nil(t, err)
+	assert.False(t, flag)
+	assert.Equal(t, config.Spec, r.ClusterConfig)
+	assert.Equal(t, int32(advCount), r.AcceptedAdvNum)
+
+	//TODO: FALSE TEST with config.MaxAcceptableAdvertisement < r.AcceptedAdvNum
```

Comment:
https://book.kubebuilder.io/reference/testing/envtest.html this may be helpful

Reply:
A simple example is the following: https://github.com/kubernetes-sigs/kubebuilder/blob/e5fda8fd51e177376e4db6c39028c2b73a7ce66e/docs/book/src/cronjob-tutorial/testdata/project/controllers/suite_test.go

Reply:
I'll check it, thanks!